contrib/rpm/skopeo.spec

```diff
@@ -0,0 +1,475 @@
+%if 0%{?fedora} || 0%{?rhel} == 6
+%global with_devel 0
+%global with_bundled 1
+%global with_debug 1
+%global with_check 0
+%global with_unit_test 0
+%else
+%global with_devel 0
+%global with_bundled 1
+%global with_debug 1
+%global with_check 0
+%global with_unit_test 0
+%endif
+
+#%%if 0%%{?with_debug}
+#%%global _dwz_low_mem_die_limit 0
+#%%else
+%global debug_package   %{nil}
+#%%endif
+
+%global provider        github
+%global provider_tld    com
+%global project         projectatomic
+%global repo            skopeo
+# https://github.com/projectatomic/skopeo
+%global provider_prefix %{provider}.%{provider_tld}/%{project}/%{repo}
+%global import_path     %{provider_prefix}
+%global git0            https://%{import_path}
+%global commit0         REPLACEWITHCOMMITID
+%global shortcommit0    %(c=%{commit0}; echo ${c:0:7})
+
+# e.g. el6 has ppc64 arch without gcc-go, so EA tag is required
+# manually listed arches due https://bugzilla.redhat.com/show_bug.cgi?id=1391932 (removed ppc64)
+ExcludeArch: ppc64
+
+Name:           %{repo}
+%if 0%{?centos}
+Epoch:          1
+%endif # centos
+Version:        0.1.27
+Release:        1.git%{shortcommit0}%{?dist}
+Summary:        Inspect Docker images and repositories on registries
+License:        ASL 2.0
+URL:            %{git0}
+Source0:        %{git0}/archive/%{commit0}/%{name}-%{shortcommit0}.tar.gz
+Source1:        storage.conf
+Source2:        containers-storage.conf.5.md
+Source3:        mounts.conf
```

Comment:
Just to reemphasize, without having these config `Source*` files available, the build will fail.

Reply:
That's similar to how buildah is now. Building the tarball from the repo is fine, and as long as the conf files are available in the repo somewhere this can be easily overcome (although, I believe mounts.conf currently isn't in the repo).

If the file is in the repo here, CI jobs can be added to validate the spec is up to date / in sync with the version details etc. when the repo versions change.

As a consumer outside Red Hat, the biggest issue I have is finding a canonical source for the spec files that Red Hat uses for its build. I can download the SRPM after the fact, but I have no mechanism or opportunity to provide feedback (or pull requests) if there are issues in the spec files that Red Hat is producing (other than bugzilla).

I am not suggesting this has to be solved here, but my view is it's practical to store it next to the source that it packages.

Reply:
> Building the tarball from the repo is fine, and as long as the conf files are available in the repo somewhere this can be easily overcome (although, I believe mounts.conf currently isn't in the repo)

None of the three are in _this_ repo.  If you are fine with getting these files from some repo different from this one, why not get the spec file from that same repo as well?

---

> As a consumer outside Red Hat, the biggest issue I have is finding a canonical source for the spec files that Red Hat uses for its build.

I’m not sure we can solve that here — **generally**, RHEL spec files are maintained internally and only published in SRPMs, or CentOS, after the relevant RHEL release is made public.  That is, AFAICT, not likely to change soon (the rationale I have quoted for making the Fedora repo, and not upstream repos, canonical for Fedora applies just as well for RHEL), so having _some_ spec file in this repo would not give you any more visibility to the internal RHEL repo before a release.

And because the RHEL spec files are not public before a release, a CI run in public GitHub cannot sync with them either.

>  I can download the SRPM after the fact, but I have no mechanism or opportunity to provide feedback (or pull requests) if there are issues in the spec files that Red Hat is producing (other than bugzilla).

Bugzilla is the _right_ place for RHEL feedback and contributions, in general.

---

Given the _general_ situation described above, there might perhaps be something specifically different about `skopeo`, due to Atomic Host and frequent RHEL Atomic releases and the like, which might give you easier access to WIP work, or something close to it, for `skopeo`.

But that would happen in the packaging efforts. Maybe @lsm5 can suggest something.

Reply:
Thanks for the response @mtrmac.

These files are currently available in this repo under the contrib directory
`storage.conf`
`containers-storage.conf.5.md`

It's only mounts.conf that is currently missing.

I am not really fine from getting them elsewhere, it would be nice to be able to understand where they are coming from, and once again be able to have a source I can provide feedback/PRs to. If I do have to get these files from somewhere else, i'd ask what the process is for determining that 'somewhere else'?

This is a more general issue with Red Hat making spec files available for public comment/commit, and I should probably raise it back through Red Hat channels.

Reply:
btw, any CentOS mentions in the spec file are RE: CentOS Virt SIG (rebuilt from rawhide) and not CentOS Base (rebuilt from RHEL).

@pixdrift so, the major change in the spec file here is renaming skopeo-containers to containers-common, yes?
While people are using the `spec_in_upstream` approach to do things like automatic COPR builds, this spec file would often end up being stale compared to the canonical dist-git spec. So, as long as you're ok with keeping this spec updated and upstream is ok with frequent PRs to it, fine by me. Issues on RHEL/CentOS packages are best reported via bugzilla like @mtrmac mentioned.

If you choose to send PRs to this repo for spec file changes, please make sure to rebase the spec file on what's in the srpm. Also, if you're a user of CentOS Virt SIG, it would be best to send PRs to the fedora rawhide dist-git for skopeo, since that's where CentOS Virt SIG is built from. HTH.

Reply:
@lsm5, what's the method I should use to get access to the canonical source and the files required to build the rpm? That is what I was hoping to solve.

The package name change is the only spec change (original intention was just to include it and the missing source files). Not sure how to raise/co-ordinate the package name upstream, if bugzilla is the method I can use that.

Reply:
@lsm5, it does help, thanks.

It feels as if the source for the rpms is 'locked away' in Red Hat, and the only way for an outsider to access them outside of Red Hat is to wait for Red Hat to produce a product.

It's a much bigger issue than this repo, i'll admit, and I appreciate that packaging is in the vendor space... but I don't think it's crazy for the project to provide / supply their own rpm spec file (or dpkg sources for that matter) as part of the project so an rpm can be built independent of any vendor.

Downloading the src rpm works but it's at the mercy of a vendor/supplier release schedule, the rpm spec from the vendor also doesn't describe where the components used to create the rpm originated from.

Rather than the rpm fall behind and need to be kept in sync with upstream, an alternative view is that it is kept in sync with the repo packaging through transparent (ie. visible) CI processes and can be included/involved in the validation process of the pipeline when version numbers / files change in the repo.

Reply:
> While people are using the spec_in_upstream approach to do things like automatic COPR builds, this spec file would often end up being stale compared to the canonical dist-git spec. So, as long as you're ok with keeping this spec updated and upstream is ok with frequent PRs to it, fine by me.

I’m fine with keeping my hands off and letting you two, or whoever is relevant, maintain the spec file in here, with as little friction as possible (e.g. @lsm5 merging PRs without my involvement), _as long as there is consensus on the benefit_.

But, given
> Also, if you're a user of CentOS Virt SIG, it would be best to send PRs to the fedora rawhide dist-git for skopeo, since that's where CentOS Virt SIG is built from.

Reply:
and
> for changes to rhel/centos base, file a bug and add your patch if any, and wait for package maintainers to ship builds with that fix.

vs.
> what's the method I should use to get access to the canonical source and the files required to build the rpm?

it is still not clear to me that this is actually solving anything: If I understand correctly, @pixdrift wants to be involved in the spec file changes during RHEL development, but this PR would not change anything about the status quo, which is that RHEL specs are only revealed after a public release of a supported package.

And the Fedora and CentOS packages are also maintained elsewhere, and @lsm5 wants to accept PRs there, not here.

So, is maintaining an _extra_ copy, which is (if I am reading the discussion correctly) _still not_ the primary place of RHEL development, making anyone’s life easier?

Reply:
@mtrmac yup, you're right. Having a spec here would mostly just be an added cost unless it's being used for something like COPR automation.